### ISSUE TITLE
CJS/ESM Build Alignment

### ISSUE THREAD

**Issue by gristow** [0xfe/vexflow#1346]
The output of the CJS and ESM builds from `src/index.ts` have different shapes because, in the Webpack UMD build process, the default export of Vex in `src/index.ts` overwrites the named exports from the same file.

As a result, when working in an ESM context this will work:
```typescript
import { StaveNote } from 'vexflow';
const sn = new StaveNote();
```

But the parallel CJS code will not work:
```typescript
const StaveNote = require('vexflow').StaveNote;
const sn = new StaveNote(); // Uncaught Reference Error: StaveNote is not defined
```

I'm hitting this issue, personally, using Vite which transforms ESM import style code to CJS require style code in node to provide the server-side render. (And I would guess given the rise in SSR frameworks we'll soon see others with similar issues.)

A quick fix would be to merge Flow onto the default export in `src/index.ts`:
```typescript
// Export Vex as a named export AND as a default export, to support both ways of importing.
export { Vex };
export default {
  ...Vex,
  ...Flow,
};
```

This alters the default export to become an object, instead of the Vex class -- but, since we only use static props on the Vex class anyway, this doesn't break any tests. (And it shouldn't break any implementations.)

Thoughts? (Especially @ronyeh since you did the bulk of the ESM build work.)

I've got a working branch with these changes -- but will hold off on a PR unless/until we think this makes sense as a way to go.

**Comment by 0xfe** [comment]
Thanks @gristow. I don't have a strong opinion here, the quick fix sounds okay to me. But @ronyeh might have other suggestions.

**Comment by ronyeh** [comment]
Sorry for the delay. I was busy this week with deadlines.

TLDR: I agree that we should strive for maximum flexibility, so that VexFlow is compatible with as many tools as possible. However, we should make sure this PR doesn't **also** change the shape of the ESM build. Ideally the ESM import and CJS require will return identical objects (if possible).

---

I left the CJS export untouched while building the ESM support. I didn't want to break any of the existing CJS code, so the shape of the CJS output is the same as it has always been. Calling `require('vexflow')` gives you the `Vex` object. To get the `StaveNote` class, you currently still need to do `Vex.Flow.StaveNote` like in VexFlow 3.0.9.

So yes, the ESM build provides much more flexibility vs the CJS build.

Here is some test code which works with the ESM build:

```javascript
// ESM Example

import { Vex } from 'vexflow';
import { StaveNote } from 'vexflow';

// Get the default export.
import VexDefault from 'vexflow';

// Dynamic import, which returns a Module object of the form:
// {
//   ...Vex,
//   Vex: Vex,
//   default: Vex
// }
const VexDynamic = await import('vexflow');

console.log(Vex.Flow.BUILD);

// All console.log(...) statements below output true.

console.log(VexDefault === Vex);
console.log(StaveNote === Vex.Flow.StaveNote);
console.log(StaveNote === VexDefault.Flow.StaveNote);

Vex.Flow.FOO = 1;
console.log(VexDefault.Flow.FOO === 1);

console.log(VexDynamic.default === VexDefault);
console.log(VexDynamic.Flow.StaveNote == StaveNote);
console.log(VexDynamic.default.Flow.StaveNote == StaveNote);

console.log(VexDynamic.Flow.FOO === 1);
console.log(VexDynamic.default.Flow.FOO === 1);

console.log(VexDynamic.Vex === VexDynamic.default);
console.log(VexDynamic.Vex.Flow === VexDynamic.default.Flow);
console.log(VexDynamic.Flow === VexDynamic.default.Flow);
```

I need to test this, but if we change `src/index.ts`, won't that affect the imported ESM object?

Perhaps we could just change the webpack CJS build somehow via a configuration setting?

https://webpack.js.org/configuration/output/#outputlibraryexport

Currently our Gruntfile specifies this webpack config for the CJS build:
```
        library: {
          name: libraryName,
          type: 'umd',
          export: 'default',
        },
```

I wonder if there is a way we can tweak the `export` setting to output the correct thing.

**Comment by ronyeh** [comment]
Take a look at PR https://github.com/0xfe/vexflow/pull/1349

I believe the approach there solves this issue, but I'll need to put together a build so you all can verify!
